### ISSUE TITLE
kube-proxy often comes up in a broken state and does not correct itself

### ISSUE THREAD

**Issue by chancez** [coreos/bootkube#111]
TL;DR: run the hack/multi-node setup and check the logs of your kube-proxy pods.

Running the vanilla, default hack/multi-node setup, I find that a large majority of the time, `kube-proxy` comes up, and cannot connect to the kube-apiserver, and never succeeds. Usually, all kube-proxies are effected, but not always. The logs are in the gist below. 

This severely breaks all functionality within the cluster, as all control components will use the `kubernetes.default` hostname or `10.3.0.1` which is the api server's cluster IP. This means the scheduler, controller-manager, etc cannot function if their node's local kube-proxy is in this broken state. As a result, this bug tends to prevent the cluster from working entirely, as the scheduler/controller-manager stop functioning. Deleting/restart the kube-proxy can resolve this as it tends to come back up in a working state, but if the scheduler was not working when deleting the kube-proxy pod, the proxy pod will not come back and the cluster remains broken.

This issue seems to have something to do with the pivot of the bootkube API server to the self-hosted API server, but I cannot say for sure. The self-hosted API server logs seem normal.

https://gist.github.com/chancez/b802af6729bfd03a6b629639fa76cd5d

I've got one idea for potentially fixing this, and that might be to configure the controller-manager/scheduler to use the external IP/hostname instead of the service, but I'm not fond of that, or sure if that would resolve it, but rather just make it so the cluster doesn't get into an unrecoverable state. Ideally, we find out why kube-proxy never fixes itself as that seems to be the root issue here.

**Comment by aaronlevy** [comment]
@chancez are you sure those timestamps in the kube-proxy are not old? During bootstrap it's normal that the kube-proxy may not be able to contact the api-server for a time, but will recover (but it doesn't say anything on default log levels, just stops logging failures). So may seem like last action is failure -- but logline is old.

Also, are you able to hit the api-server from the host directly? The kube-proxy uses host-network, so if there are connectivity problems there, they should also surface directly from the host.

It may help to bump up the logging level of the proxy - as the information in that gist doesn't tell us much.

**Comment by chancez** [comment]
@aaronlevy I'm sure they aren't old, they continue to error after bootstrap, forever after bring up until the pod is recreated.

The proxy is reachable from outside the cluster, as I am using kubectl logs to view my proxy logs.  If I ssh onto any node I can reach the API from inside the VM using the public address of the API (https://172.17.4.101:443/api in this case).

**Comment by aaronlevy** [comment]
What about if you exec in the proxy container itself? Possible network configuration of pod has something wrong.

**Comment by chancez** [comment]
Okay, turns out the logs do stop coming in from the proxy, and it seems to resolve itself. I'm going to attempt to reproduce outside of the Vagrant cluster where I was originally experiencing the issue and come back to the Vagrant example if I can.

**Comment by chancez** [comment]
I've gotten to a better point where I think kube-proxy is not the root cause in our setup. But I am still experiencing an issue which causes the cluster to get locked up and unable to progress (scheduler becomes unable to talk to the API, and then the cluster never recovers, sometimes its the controller-manager and scheduler, etc). The issue is proving extremely difficult to debug or track down so I'll reopen/create a new issue when I have more information on the issue.
